src/execution_plan/execution_plan.c

```diff
@@ -527,21 +560,150 @@ static inline void _buildUnwindOp(ExecutionPlan *plan, const cypher_astnode_t *c
 	_ExecutionPlan_UpdateRoot(plan, op);
 }
 
-static inline void _buildMergeOp(GraphContext *gc, AST *ast, ExecutionPlan *plan,
-								 const cypher_astnode_t *clause, ResultSetStatistics *stats) {
-	// A merge clause provides a single path that must exist or be created.
-	// As with paths in a MATCH query, build the appropriate traversal operations
-	// and append them to the set of ops.
-	AST_MergeContext merge_ast_ctx = AST_PrepareMergeOp(gc, clause, plan->query_graph);
-	OpBase *op = NewMergeOp(plan, stats, merge_ast_ctx.nodes_to_merge, merge_ast_ctx.edges_to_merge);
-	_ExecutionPlan_UpdateRoot(plan, op);
+static void _buildMergeMatchStream(ExecutionPlan *plan, const cypher_astnode_t *clause,
+								   const char **arguments) {
+	AST *ast = QueryCtx_GetAST();
+	// Initialize an ExecutionPlan that shares this plan's Record mapping.
+	ExecutionPlan *rhs_plan = _NewEmptyExecutionPlan();
+	rhs_plan->record_map = plan->record_map;
+
+	// If we have bound variables, build an Argument op that represents them.
+	if(arguments) rhs_plan->root = NewArgumentOp(plan, arguments);
+
+	// Build a temporary AST holding the MERGE path within a MATCH clause.
+	const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
+	AST *rhs_ast = AST_MockMatchPattern(ast, path);
```

Comment:
I generally don't like making function stacks that unfold to look like
`NewExecutionPlan -> _PopulateExecutionPlan [...] -> _buildMergeMatchStream -> NewExecutionPlan -> ...`
And feel like this would cause additional complications, because we can't use `NewExecutionPlan`'s call to `AST_Free`, etc. 

I'm open to discussing it, though!